### PR TITLE
fix(encoder,parser): escape and unescape strings 

### PR DESCRIPTION
**Pull request by goto-bus-stop**
This PR improves the string value printing in apollo-encoder. It tries to sensibly decide between a multi-line block string, a single-line block string, and a non-block string depending on the contents it needs to print. It escapes special characters when writing non-block strings. When writing block strings, only `"""` is escaped as `\"""` so the block isn't ended prematurely.

The parser's `String::from(StringValue)` implementation also un-escapes characters.

### CODE REVIEW

**Comment by goto-bus-stop** [comment]
I'm not really happy with this but I think I should revisit later rather than spend more time on it now. I'd like to more or less do the string parsing from scratch and also address #319 at that time. Atleast this should improve the string printing in apollo-encoder so it doesn't produce invalid output anymore.

File: crates/apollo-encoder/src/field.rs [hunk 1]
```diff
@@ -332,9 +332,7 @@ mod tests {
             r#"  spaceCat(
     "The type of treats given in space"
     treat: Treat,
-    """
-    Optimal age of a "space" cat
-    """
+    """Optimal age of a "space" cat"""
```
Comment (lrlna) [inline crates/apollo-encoder/src/field.rs:335]: so i think we should always format block string values with leading and trailing new line. There is a bit on that in [spec here](https://spec.graphql.org/October2021/#sel-FAFdPTDPAADVCA5ke)

**Review by lrlna — CHANGES_REQUESTED** [review]
a comment on formatting block string values. otherwise looks good!

yea, string parsing needs a brand new coat of paint. will probably be easier to solve multi-error lexing issue too.

File: crates/apollo-encoder/src/field.rs (hunk 1 above)
Comment (goto-bus-stop) [reply crates/apollo-encoder/src/field.rs]: i took this idea from graphql-js, i kinda like it for short single-line strings. but i also like having consistent output style in the encoder 🤔

Comment (lrlna) [reply crates/apollo-encoder/src/field.rs]: iiiiiinteresting they write _one_ thing in the spec, but do something _else_

Comment (lrlna) [reply crates/apollo-encoder/src/field.rs]: but sure, sounds good, let's go with what graphql-js does then

**Review by lrlna — APPROVED** [review]

File: crates/apollo-encoder/src/field.rs (hunk 1 above)
Comment (goto-bus-stop) [reply crates/apollo-encoder/src/field.rs]: i read the spec example as only a style recommendation for multi-line strings. for single-line strings you could do something else in theory 😄 

i removed it now (it's also simpler to not do it). There is one reason to do it i think: we decide whether to print a string as a block string or a non-block string based on its contents, and for single-line strings that include " but no other escapable characters, i think it does look nicer to use a single-line block string. but we can add that back if we add configurable readability things in the future.

**Comment by goto-bus-stop** [comment]
the parser unescaping implementation that i have here does the wrong thing for block strings btw. As block strings don't actually support escapes except for `\""""`. I'd improve that in the future string improvements PR.

**Pull request merged**